cloud-controller-manager/do/firewall_controller.go

```diff
@@ -156,20 +171,52 @@ func NewFirewallController(
 	return fc
 }
 
+// obj could be an *v1.Service, or a DeletionFinalStateUnknown marker item.
+func (fc *FirewallController) enqueueService(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+	fc.queue.Add(key)
+}
+
 // Run starts the firewall controller loop.
-func (fc *FirewallController) Run(ctx context.Context, stopCh <-chan struct{}, fwReconcileFrequency time.Duration) {
-	wait.Until(func() {
-		err := fc.observeRunLoopDuration(ctx, stopCh, fwReconcileFrequency)
-		if err != nil {
-			klog.Errorf("failed to run firewall controller loop: %v", err)
-		}
-	}, fwReconcileFrequency, stopCh)
+func (fc *FirewallController) Run(threadiness int, stopCh <-chan struct{}, fwReconcileFrequency time.Duration) {
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(fc.runWorker, fwReconcileFrequency, stopCh)
+	}
+
+	<-stopCh
+	klog.Info("Stopping firewall controller")
+}
+
+func (fc *FirewallController) runWorker() {
+	for fc.processNextItem() {
+	}
 }
 
-func (fc *FirewallController) actualRun(stopCh <-chan struct{}, fwReconcileFrequency time.Duration) error {
-	ctx, cancel := context.WithTimeout(context.Background(), fwReconcileFrequency)
+func (fc *FirewallController) processNextItem() bool {
+	ctx, cancel := context.WithTimeout(context.Background(), firewallReconcileFrequency)
 	defer cancel()
+	// Wait until there is a new item in the working queue
+	key, quit := fc.queue.Get()
+	if quit {
+		return false
+	}
+	// Tell the queue that we are done with processing this key. This unblocks the key for other workers
+	// This allows safe parallel processing because items with the same key are never processed in
+	// parallel.
+	defer fc.queue.Done(key)
+
+	err := fc.observeRunLoopDuration(ctx)
```

Comment:
I added a commit to move things around @timoreimann , I hope it makes sense. It ended up being a little weird because of my reliance on adding the key when checking that error.